{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [],
   "source": [
    "import polars as pl \n",
    "import numpy as np\n",
    "cov = 0.50001\n",
    "\n",
    "\n",
    "elongates = pl.read_csv(f\"output/{cov}/{cov}_elongates.csv\", infer_schema_length =10000)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "from utils.handle_UTRs import translate_frames\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import numpy as np\n",
    "\n",
    "def custom_target_elongate(cluster, scer_length, seq_id, specie, elongate_length, side, genome_dict, gff_dict, coeff = 1): \n",
    "\n",
    "\n",
    "    coordinates = list()\n",
    "    result_dict = dict()\n",
    "\n",
    "    if side != \"Nter\" and side != \"Cter\":\n",
    "\n",
    "        raise ValueError(\"Side must be either Nter or Cter\")\n",
    "    \n",
    "    gff = gff_dict[specie].filter(\n",
    "\n",
    "        (pl.col(\"Type\") == \"CDS\") & ((pl.col(\"Name\") == seq_id) | (pl.col(\"Parent\") == seq_id))\n",
    "\n",
    "    )[[\"Start\",\"End\",\"Strand\",\"Seqid\"]] # Keep only necessary columns\n",
    "    \n",
    "    # Store datas necessary to compute the elongate sequence\n",
    "\n",
    "    strand = gff[0][\"Strand\"].to_list()[0] # + or -\n",
    "    strand_id = gff[0][\"Seqid\"].to_list()[0] # chromosome or scaffold id\n",
    "\n",
    "    for row in gff.iter_rows(named=True): # Named = True to iter with column names\n",
    "\n",
    "        coordinates.append(sorted((int(row['Start'])-1, int(row['End'])-1))) # -1 for python indexing\n",
    "        \n",
    "    coordinates = sorted(coordinates, key=lambda x: x[0]) # Sort coordinates by start position\n",
    "\n",
    "\n",
    "    custom_elongate_length = int(np.ceil((scer_length - elongate_length) * 3 * coeff)) # Get the length of the elongate sequence in nucleotides\n",
    "\n",
    "    if strand == \"+\":\n",
    "\n",
    "        if side == \"Nter\":\n",
    "\n",
    "            start_5 = coordinates[0][0]-custom_elongate_length if coordinates[0][0]-custom_elongate_length >= 0 else 0 # Get the start position of the 5' UTR\n",
    "            \n",
    "            elongate = genome_dict[specie][strand_id][\"seq\"][\n",
    "\n",
    "                start_5:coordinates[0][0]\n",
    "\n",
    "            ] \n",
    "\n",
    "        elif side == \"Cter\": # Useless check but it's for the sake of clarity\n",
    "        \n",
    "            end_3 = coordinates[-1][1]+1+custom_elongate_length if coordinates[-1][1]+1+custom_elongate_length <= genome_dict[specie][strand_id][\"len\"] else genome_dict[specie][strand_id][\"len\"] # Get the end position of the 3' UTR\n",
    "            # +1 for -1,1 because GFF points to the last nucleotide of the stop codon\n",
    "\n",
    "            elongate = genome_dict[specie][strand_id][\"seq\"][\n",
    "                coordinates[-1][1]+1:end_3\n",
    "                ] # Get the 3' sequence\n",
    "\n",
    "    \n",
    "    # Reverse complement if the strand is negative, don't forget to reverse the coordinates\n",
    "    if strand == \"-\":\n",
    "\n",
    "        if side == \"Nter\":\n",
    "\n",
    "            end_5 = coordinates[-1][1]+1+custom_elongate_length if coordinates[-1][1]+1+custom_elongate_length <= genome_dict[specie][strand_id][\"len\"] else genome_dict[specie][strand_id][\"len\"] # Get the start position of the 5' UTR\n",
    "\n",
    "            elongate = genome_dict[specie][strand_id][\"seq\"][\n",
    "                coordinates[-1][1]+1:end_5\n",
    "            ].reverse_complement() # Get the 5' sequence\n",
    "        \n",
    "\n",
    "        if side == \"Cter\":\n",
    "        \n",
    "            start_3 = coordinates[0][0]-custom_elongate_length if coordinates[0][0]-custom_elongate_length >= 0 else 0 # Get the end position of the 3' UTR\n",
    "\n",
    "            elongate = genome_dict[specie][strand_id][\"seq\"][\n",
    "                start_3:coordinates[0][0]\n",
    "            ].reverse_complement() # Get the 3' sequence\n",
    "\n",
    "\n",
    "    result_dict[\"nuc\"] = SeqRecord(seq = Seq(elongate), id = f\"{seq_id}\", description = \"\") \n",
    "\n",
    "    # def translate_frames(dna_sequence, specie, seq_id, length, utr, cluster)\n",
    "\n",
    "    result_dict[\"prot\"] = translate_frames(dna_sequence = elongate, specie = specie, seq_id = seq_id, length = custom_elongate_length, utr = side, cluster = cluster)\n",
    "\n",
    "    return result_dict\n",
    "    \n",
    "    \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "def custom_scer_elongates(scer_id : str, specie : str, scer_length : int, scer_elongate : str or Seq, infos_dict, cluster, gff_dict, genome_dict, side):\n",
    "\n",
    "    coordinates = list()\n",
    "    result_dict = dict()\n",
    "\n",
    "\n",
    "    ########\n",
    "    # Get infos from gff file \n",
    "    ########\n",
    "\n",
    "    if side != \"Nter\" and side != \"Cter\":\n",
    "\n",
    "        raise ValueError(\"Side must be either Nter or Cter\")\n",
    "    \n",
    "    gff = gff_dict[specie].filter(\n",
    "\n",
    "        (pl.col(\"Type\") == \"CDS\") & ((pl.col(\"Name\") == scer_id) | (pl.col(\"Parent\") == scer_id))\n",
    "\n",
    "    )[[\"Start\",\"End\",\"Strand\",\"Seqid\"]] # Keep only necessary columns\n",
    "    \n",
    "    # Store datas necessary to compute the elongate sequence\n",
    "\n",
    "    strand = gff[0][\"Strand\"].to_list()[0] # + or -\n",
    "    strand_id = gff[0][\"Seqid\"].to_list()[0] # chromosome or scaffold id\n",
    "\n",
    "    for row in gff.iter_rows(named=True): \n",
    "\n",
    "        coordinates.append(sorted((int(row['Start'])-1, int(row['End'])-1))) # -1 for python indexing\n",
    "        \n",
    "    coordinates = sorted(coordinates, key=lambda x: x[0]) # Sort coordinates by start position\n",
    "\n",
    "    ########\n",
    "    # Check that input data are correct\n",
    "    ########\n",
    "\n",
    "    scer_elongate = scer_elongate.replace(\"-\",\"\")\n",
    "\n",
    "    if len(scer_elongate) != scer_length:\n",
    "\n",
    "        print(f\"Seq id : {scer_id}\")\n",
    "        print(f\"Cluster : {cluster}\")\n",
    "        raise ValueError(\"Mismatch between theorical and real length of the Scer peptidic elongate\")\n",
    "\n",
    "    for subject_id, subject_length in infos_dict.items():\n",
    "\n",
    "        result_dict[subject_id] = dict()\n",
    "        custom_length = (scer_length-subject_length)*3\n",
    "\n",
    "        if strand == \"+\":\n",
    "\n",
    "            if side == \"Nter\":\n",
    "                \n",
    "               nuc_elongate = genome_dict[specie][strand_id][\"seq\"][\n",
    "\n",
    "                    coordinates[0][0]:coordinates[0][0] + custom_length # Get the 5' sequence\n",
    "\n",
    "                ] \n",
    "\n",
    "            elif side == \"Cter\": # Useless check but it's for the sake of clarity\n",
    "            \n",
    "\n",
    "                nuc_elongate = genome_dict[specie][strand_id][\"seq\"][\n",
    "                    \n",
    "                    coordinates[-1][1]-(custom_length):coordinates[-1][1]] # Get the 3' sequence\n",
    "\n",
    "    \n",
    "        # Reverse complement if the strand is negative, don't forget to reverse the coordinates\n",
    "        if strand == \"-\":\n",
    "\n",
    "            if side == \"Nter\":\n",
    "\n",
    "\n",
    "                nuc_elongate = genome_dict[specie][strand_id][\"seq\"][\n",
    "                    coordinates[-1][1]-(custom_length):coordinates[-1][1]\n",
    "                ].reverse_complement() # Get the 5' sequence\n",
    "            \n",
    "\n",
    "            elif side == \"Cter\":\n",
    "            \n",
    "                nuc_elongate = genome_dict[specie][strand_id][\"seq\"][\n",
    "                    coordinates[0][0]:coordinates[0][0] + (custom_length)\n",
    "                ].reverse_complement() # Get the 3' sequence\n",
    "\n",
    "\n",
    "\n",
    "        result_dict[subject_id][\"nuc\"] = SeqRecord( seq = Seq(nuc_elongate), id = f\"{scer_id}\", description = \"\")\n",
    "        result_dict[subject_id][\"prot\"] = SeqRecord( seq = Seq(scer_elongate[:int(custom_length/3)]), id = f\"{scer_id}\", description = \"\")\n",
    "    \n",
    "    return result_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [],
   "source": [
    "import yaml\n",
    "# Load your species regex from the yaml file\n",
    "with open('/home/simon.herman/Bureau/Gits/Elongates/env.yaml', 'r') as f:\n",
    "    yaml_data = yaml.safe_load(f)\n",
    "    species_order = yaml_data['Species_order']['Scer']\n",
    "\n",
    "seuil = 4\n",
    "\n",
    "Nter_scer_conditions = ((pl.col(\"species\") == \"Scer_NCBI\") & (abs(pl.col(\"max_Nter\") - pl.col(\"Nter_nb_aa\")) < seuil) & (pl.col(\"max_Nter\") >= 10))\n",
    "Cter_scer_conditions = ((pl.col(\"species\") == \"Scer_NCBI\") & (abs(pl.col(\"max_Cter\") - pl.col(\"Cter_nb_aa\")) < seuil) & (pl.col(\"max_Cter\") >= 10))\n",
    "test = elongates.filter(Nter_scer_conditions | Cter_scer_conditions)[\"cluster_name\"].unique().to_list()\n",
    "\n",
    "nter_clusters = []\n",
    "cter_clusters = []\n",
    "i = 0\n",
    "\n",
    "for cluster,sequences in elongates.filter(pl.col(\"cluster_name\").is_in(test)).groupby(\"cluster_name\"):\n",
    "\n",
    "    species = sequences[\"species\"].to_list()\n",
    "\n",
    "    if len(species) < 3:\n",
    "            \n",
    "        continue\n",
    "\n",
    "\n",
    "    species_ordered = [ s for s in species_order if s in species]\n",
    "    \n",
    "    if species_ordered[-1] != \"Scer_NCBI\" and species.count(species_ordered[-1]) > 1:\n",
    "            \n",
    "        continue\n",
    "\n",
    "    if sequences.filter(pl.col(\"species\") == species_ordered[-1])[\"Nter_nb_aa\"].to_list()[0] < seuil and sequences[\"max_Nter\"].max() >= 10:\n",
    "    \n",
    "        nter_clusters.append(cluster)\n",
    "\n",
    "    if sequences.filter(pl.col(\"species\") == species_ordered[-1])[\"Cter_nb_aa\"].to_list()[0] < seuil and sequences[\"max_Cter\"].max() >= 10:\n",
    "    \n",
    "        cter_clusters.append(cluster)\n",
    "\n",
    "\n",
    "\n",
    "full_filtered_Nter = elongates.filter(pl.col(\"cluster_name\").is_in(nter_clusters))\n",
    "full_filtered_Cter = elongates.filter(pl.col(\"cluster_name\").is_in(cter_clusters))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gff3_parser\n",
    "from utils.files import multifasta_to_dict\n",
    "gff_dict = dict()\n",
    "genome_dict = dict()\n",
    "\n",
    "for specie in species_order:\n",
    "\n",
    "    gff_dict[specie] = pl.from_pandas(gff3_parser.parse_gff3(f\"input/{specie}.gff\", parse_attributes = True, verbose = False))\n",
    "    genome_dict[specie] = multifasta_to_dict(f\"input/{specie}.fna\", genome = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "import shutil\n",
    "from Bio import SeqIO\n",
    "\n",
    "current_path = f\"/home/simon.herman/Bureau/Gits/Elongates/work/{cov}\"\n",
    "\n",
    "\n",
    "\n",
    "dataframes = {\n",
    "    \"Nter\": full_filtered_Nter,\n",
    "    \"Cter\": full_filtered_Cter,\n",
    "}\n",
    "\n",
    "if os.path.exists(f\"{current_path}/local_align_files\"):\n",
    "\n",
    "    shutil.rmtree(f\"{current_path}/local_align_files\")\n",
    "\n",
    "os.mkdir(f\"{current_path}/local_align_files\")\n",
    "\n",
    "for side in [\"Nter\", \"Cter\"]:\n",
    "\n",
    "    df = dataframes[side]\n",
    "\n",
    "    os.mkdir(f\"{current_path}/local_align_files/{side}\")\n",
    "\n",
    "    for cluster, sequences in df.groupby(\"cluster_name\"):\n",
    "\n",
    "        scer_length = sequences.filter(pl.col(\"species\") == \"Scer_NCBI\")[f\"{side}_nb_aa\"].max() # Maybe several Scer sequences in the cluster, we take the longest elongate\n",
    "        \n",
    "        if scer_length < 10:\n",
    "            continue\n",
    "\n",
    "        scer_id = sequences.filter((pl.col(\"species\") == \"Scer_NCBI\") & (pl.col(f\"{side}_nb_aa\") == scer_length))[\"seq_id\"].to_list()[0]\n",
    "        os.mkdir(f\"{current_path}/local_align_files/{side}/{cluster}\")\n",
    "     \n",
    "        infos_for_scer_sequences = dict()\n",
    "        for sequence in sequences.iter_rows(named = True): \n",
    "\n",
    "            if scer_length - sequence[f\"{side}_nb_aa\"] >= 10:\n",
    "\n",
    "\n",
    "                os.mkdir(f\"{current_path}/local_align_files/{side}/{cluster}/{sequence['seq_id']}\")\n",
    "                os.mkdir(f\"{current_path}/local_align_files/{side}/{cluster}/{sequence['seq_id']}/nucleotide\")\n",
    "                os.mkdir(f\"{current_path}/local_align_files/{side}/{cluster}/{sequence['seq_id']}/protein\")\n",
    "\n",
    "\n",
    "                dict_ = custom_target_elongate(cluster, scer_length, sequence[\"seq_id\"], sequence[\"species\"], sequence[f\"{side}_nb_aa\"], side, genome_dict, gff_dict)\n",
    "\n",
    "            \n",
    "                \n",
    "                SeqIO.write(dict_[\"nuc\"], f\"{current_path}/local_align_files/{side}/{cluster}/{sequence['seq_id']}/nucleotide/{sequence['seq_id']}.fna\", \"fasta\")\n",
    "                \n",
    "                SeqIO.write([ seq for seq in dict_[\"prot\"].values()], f\"{current_path}/local_align_files/{side}/{cluster}/{sequence['seq_id']}/protein/{sequence['seq_id']}.faa\", \"fasta\")\n",
    "                \n",
    "                infos_for_scer_sequences[sequence[\"seq_id\"]] = sequence[f\"{side}_nb_aa\"] \n",
    "                \n",
    "        \n",
    "        scer_cds = sequences.filter(pl.col(\"seq_id\") == scer_id)\n",
    "        \n",
    "        scer_elongates = custom_scer_elongates(scer_id = scer_id, specie = \"Scer_NCBI\", scer_length = scer_length, \n",
    "                                                    scer_elongate = scer_cds[f\"{side}_elongate\"].to_list()[0], infos_dict = infos_for_scer_sequences, \n",
    "                                                    cluster = cluster, gff_dict = gff_dict, genome_dict = genome_dict, side = side)\n",
    "\n",
    "\n",
    "        for subject_sequence in scer_elongates.keys():\n",
    "\n",
    "            SeqIO.write(scer_elongates[subject_sequence][\"nuc\"], f\"{current_path}/local_align_files/{side}/{cluster}/{subject_sequence}/nucleotide/{scer_id}_custom.fna\", \"fasta\")\n",
    "            SeqIO.write(scer_elongates[subject_sequence][\"prot\"], f\"{current_path}/local_align_files/{side}/{cluster}/{subject_sequence}/protein/{scer_id}_custom.faa\", \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import polars as pl\n",
    "from polars import NoDataError\n",
    "from Bio import SeqIO\n",
    "import re\n",
    "\n",
    "# Fields: query id, subject id, % identity, alignment length, mismatches, gap opens, q. start, q. end, s. start, s. end, evalue, bit score\n",
    "dtypes = [pl.Utf8, pl.Utf8, pl.Float64, pl.Int64, pl.Int64, pl.Int64, pl.Int64, pl.Int64, pl.Int64, pl.Int64, pl.Float64, pl.Float64]\n",
    "dfs = list()\n",
    "trunc_files = open(f\"output/{cov}/trunc_files.txt\", \"a\")\n",
    "current_path = f\"/home/simon.herman/Bureau/Gits/Elongates/work/{cov}/local_align_files\"\n",
    "i = 0\n",
    "for side in os.listdir(current_path):\n",
    "\n",
    "    for cluster in os.listdir(f\"{current_path}/{side}\"):\n",
    "\n",
    "        for subject_seq in os.listdir(f\"{current_path}/{side}/{cluster}\"):\n",
    "\n",
    "            for condition in [\"nucleotide\",\"protein\"]:\n",
    "\n",
    "                root = f\"{current_path}/{side}/{cluster}/{subject_seq}/{condition}/\"\n",
    "                if os.path.exists(output_csv):\n",
    "                    os.remove(output_csv)\n",
    "\n",
    "                to_align = os.listdir(f\"{current_path}/{side}/{cluster}/{subject_seq}/{condition}\")\n",
    "\n",
    "                query_index = next((i for i, item in enumerate(to_align) if \"custom\" in item))\n",
    "                subject_index = not query_index # List length 2 in every scenario\n",
    "\n",
    "                query = os.path.join(root, to_align[query_index])\n",
    "                subject = os.path.join(root, to_align[subject_index])      \n",
    "                output_csv = os.path.join(root,\"temp.tsv\")      \n",
    "                \n",
    "                if condition == \"nucleotide\":\n",
    "                    command = f\"lalign36 -3 -E 1000 -m 8 {query} {subject} > {output_csv}\"\n",
    "                elif condition == \"protein\":\n",
    "                    command = f\"lalign36 -p -3 -E 1000 -m 8 {query} {subject} > {output_csv}\"\n",
    "                \n",
    "                subprocess.run(command, shell=True)\n",
    "                try:\n",
    "                    df = pl.read_csv(output_csv, separator=\"\\t\", has_header=False, dtypes = dtypes)\n",
    "                except:\n",
    "                    trunc_files.write(f\"{query}\\n{subject_index}\\n----------------\\n\")\n",
    "                    NoDataError()\n",
    "\n",
    "                if condition == \"nucleotide\":\n",
    "                    df = df.sort(\"column_11\", descending = False).head(1).with_columns(\n",
    "                        Condition = pl.lit(condition),\n",
    "                        Side = pl.lit(side))\n",
    "\n",
    "                if condition == \"protein\":\n",
    "                    tmp_ = []\n",
    "                    for frame, align_res in df.groupby(\"column_2\"):\n",
    "                        tmp_.append(align_res.sort(\"column_11\", descending = False).head(1).with_columns(\n",
    "                            Condition = pl.lit(condition),\n",
    "                            Side = pl.lit(side)))\n",
    "                    df = pl.concat(tmp_)\n",
    "\n",
    "                dfs.append(df)\n",
    "                os.remove(output_csv)\n",
    "                i += 1\n",
    "                \n",
    "        if i > 20:\n",
    "            break\n",
    "\n",
    "    if i > 20:\n",
    "        break\n",
    "\n",
    "\n",
    "                                    \n",
    "                \n",
    "df = pl.concat(dfs)\n",
    "trunc_files.close()\n",
    "\n",
    "\n",
    "df = pl.concat(dfs)\n",
    "df = df.with_columns(\n",
    "\n",
    "    pl.lit(\"NA\").alias('subject'),\n",
    "    pl.lit(\"NA\").alias('relative_frame'),\n",
    "    pl.lit(\"NA\").alias('q_specie'),\n",
    "    pl.lit(\"NA\").alias('s_specie')\n",
    ")\n",
    "\n",
    "\n",
    "                \n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['rna-XM_033908579.1-cluster_n3567-Nter-f2-53',\n",
       " 'rna-XM_033908579.1-cluster_n3567-Nter-f1-53',\n",
       " 'rna-XM_033908579.1-cluster_n3567-Nter-f0-53',\n",
       " 'rna-XM_033908579.1-cluster_n3567-Nter-f2-53',\n",
       " 'Sarb_01G00560.1-cluster_n3567-Nter-f2-562',\n",
       " 'Sarb_01G00560.1-cluster_n3567-Nter-f1-562',\n",
       " 'Sarb_01G00560.1-cluster_n3567-Nter-f0-562',\n",
       " 'Sarb_01G00560.1-cluster_n3567-Nter-f2-562',\n",
       " 'rna-XM_033911492.1-cluster_n4261-Nter-f2-10',\n",
       " 'rna-XM_033911492.1-cluster_n4261-Nter-f1-10',\n",
       " 'rna-XM_033911492.1-cluster_n4261-Nter-f2-10',\n",
       " 'rna-XM_033911492.1-cluster_n4261-Nter-f0-10',\n",
       " 'Sbay_12.217-cluster_n4261-Nter-f2-24',\n",
       " 'Sbay_12.217-cluster_n4261-Nter-f0-24',\n",
       " 'Sbay_12.217-cluster_n4261-Nter-f2-24',\n",
       " 'Sbay_12.217-cluster_n4261-Nter-f1-24',\n",
       " 'Smik_10.419-cluster_n4261-Nter-f2-24',\n",
       " 'Smik_10.419-cluster_n4261-Nter-f0-24',\n",
       " 'Smik_10.419-cluster_n4261-Nter-f2-24',\n",
       " 'Smik_10.419-cluster_n4261-Nter-f1-24',\n",
       " 'Skud_10.350-cluster_n4261-Nter-f2-24',\n",
       " 'Skud_10.350-cluster_n4261-Nter-f0-24',\n",
       " 'Skud_10.350-cluster_n4261-Nter-f2-24',\n",
       " 'Skud_10.350-cluster_n4261-Nter-f1-24',\n",
       " 'Smik_16.204-cluster_n1942-Nter-f2-56',\n",
       " 'Smik_16.204-cluster_n1942-Nter-f2-56',\n",
       " 'Smik_16.204-cluster_n1942-Nter-f1-56',\n",
       " 'Smik_16.204-cluster_n1942-Nter-f0-56',\n",
       " 'Sarb_05G02360.1-cluster_n108-Nter-f2-111',\n",
       " 'Sarb_05G02360.1-cluster_n108-Nter-f0-111',\n",
       " 'Sarb_05G02360.1-cluster_n108-Nter-f2-111',\n",
       " 'Sarb_05G02360.1-cluster_n108-Nter-f1-111',\n",
       " 'Sbay_8.443-cluster_n6748-Nter-f2-216',\n",
       " 'Sbay_8.443-cluster_n6748-Nter-f1-216',\n",
       " 'Sbay_8.443-cluster_n6748-Nter-f0-216',\n",
       " 'Sbay_8.443-cluster_n6748-Nter-f2-216',\n",
       " 'rna-XM_033909638.1-cluster_n1667-Nter-f2-64',\n",
       " 'rna-XM_033909638.1-cluster_n1667-Nter-f2-64',\n",
       " 'rna-XM_033909638.1-cluster_n1667-Nter-f0-64',\n",
       " 'rna-XM_033909638.1-cluster_n1667-Nter-f1-64',\n",
       " 'Sarb_13G03880.1-cluster_n1667-Nter-f2-64',\n",
       " 'Sarb_13G03880.1-cluster_n1667-Nter-f1-64',\n",
       " 'Sarb_13G03880.1-cluster_n1667-Nter-f0-64',\n",
       " 'Sarb_13G03880.1-cluster_n1667-Nter-f2-64']"
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.select(\"column_2\").to_series().to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_blast_dataframe(row : tuple) -> tuple:\n",
    "\n",
    "\n",
    "\n",
    "    ###################\n",
    "\n",
    "\n",
    "    # MANQUE A PARSER LES SEQ ID DU DATAFRAME EN FONCTION DE SI CEST NUC OU PROT : FAIRE DEUX FONCTIONS POUR APPLY ET ENSUITE GROUPBY\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    prot_subject_pattern = r'(.*?)-(cluster_n\\d+)-(.*)-(f\\d+)-(\\d+)'\n",
    "\n",
    "    query_id = row[0]\n",
    "    subject_id = row[1]\n",
    "    identity = row[2]\n",
    "    align_lenght = row[3]\n",
    "    mismatches = row[4]\n",
    "    gap_opens = row[5]\n",
    "    qstart = row[6]\n",
    "    qend = row[7]\n",
    "    sstart = row[8]\n",
    "    send = row[9]\n",
    "    evalue = row[10]\n",
    "    bitscore = row[11]\n",
    "\n",
    "    match = re.search(prot_subject_pattern, subject_id)\n",
    "\n",
    "\n",
    "    tuple_ = tuple([query_seq_id, subject_seq_id, evalue, qstart, qend, sstart, send, \n",
    "                    qseq, sseq, length, blast_gapopen, blast_gaps, query_cluster, subject_cluster, subject_relative_frame, \n",
    "                    is_same_cluster, q_specie, s_specie])\n",
    "    return tuple_"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
