{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import polars as pl \n",
    "import numpy as np\n",
    "cov = 0.50001\n",
    "\n",
    "\n",
    "elongates = pl.read_csv(f\"output/{cov}/{cov}_elongates.csv\", infer_schema_length =10000)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "from utils.handle_UTRs import translate_frames\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import numpy as np\n",
    "\n",
    "def custom_target_elongate(cluster, scer_length, seq_id, specie, elongate_length, side, genome_dict, gff_dict, coeff = 1): \n",
    "\n",
    "\n",
    "    coordinates = list()\n",
    "    result_dict = dict()\n",
    "\n",
    "    if side != \"Nter\" and side != \"Cter\":\n",
    "\n",
    "        raise ValueError(\"Side must be either Nter or Cter\")\n",
    "    \n",
    "    gff = gff_dict[specie].filter(\n",
    "\n",
    "        (pl.col(\"Type\") == \"CDS\") & ((pl.col(\"Name\") == seq_id) | (pl.col(\"Parent\") == seq_id))\n",
    "\n",
    "    )[[\"Start\",\"End\",\"Strand\",\"Seqid\"]] # Keep only necessary columns\n",
    "    \n",
    "    # Store datas necessary to compute the elongate sequence\n",
    "\n",
    "    strand = gff[0][\"Strand\"].to_list()[0] # + or -\n",
    "    strand_id = gff[0][\"Seqid\"].to_list()[0] # chromosome or scaffold id\n",
    "\n",
    "    for row in gff.iter_rows(named=True): # Named = True to iter with column names\n",
    "\n",
    "        coordinates.append(sorted((int(row['Start'])-1, int(row['End'])-1))) # -1 for python indexing\n",
    "        \n",
    "    coordinates = sorted(coordinates, key=lambda x: x[0]) # Sort coordinates by start position\n",
    "\n",
    "\n",
    "    custom_elongate_length = int(np.ceil((scer_length - elongate_length) * 3 * coeff)) # Get the length of the elongate sequence in nucleotides\n",
    "\n",
    "    if strand == \"+\":\n",
    "\n",
    "        if side == \"Nter\":\n",
    "\n",
    "            start_5 = coordinates[0][0]-custom_elongate_length if coordinates[0][0]-custom_elongate_length >= 0 else 0 # Get the start position of the 5' UTR\n",
    "            \n",
    "            elongate = genome_dict[specie][strand_id][\"seq\"][\n",
    "\n",
    "                start_5:coordinates[0][0]\n",
    "\n",
    "            ] \n",
    "\n",
    "        elif side == \"Cter\": # Useless check but it's for the sake of clarity\n",
    "        \n",
    "            end_3 = coordinates[-1][1]+1+custom_elongate_length if coordinates[-1][1]+1+custom_elongate_length <= genome_dict[specie][strand_id][\"len\"] else genome_dict[specie][strand_id][\"len\"] # Get the end position of the 3' UTR\n",
    "            # +1 for -1,1 because GFF points to the last nucleotide of the stop codon\n",
    "\n",
    "            elongate = genome_dict[specie][strand_id][\"seq\"][\n",
    "                coordinates[-1][1]+1:end_3\n",
    "                ] # Get the 3' sequence\n",
    "\n",
    "    \n",
    "    # Reverse complement if the strand is negative, don't forget to reverse the coordinates\n",
    "    if strand == \"-\":\n",
    "\n",
    "        if side == \"Nter\":\n",
    "\n",
    "            end_5 = coordinates[-1][1]+1+custom_elongate_length if coordinates[-1][1]+1+custom_elongate_length <= genome_dict[specie][strand_id][\"len\"] else genome_dict[specie][strand_id][\"len\"] # Get the start position of the 5' UTR\n",
    "\n",
    "            elongate = genome_dict[specie][strand_id][\"seq\"][\n",
    "                coordinates[-1][1]+1:end_5\n",
    "            ].reverse_complement() # Get the 5' sequence\n",
    "        \n",
    "\n",
    "        if side == \"Cter\":\n",
    "        \n",
    "            start_3 = coordinates[0][0]-custom_elongate_length if coordinates[0][0]-custom_elongate_length >= 0 else 0 # Get the end position of the 3' UTR\n",
    "\n",
    "            elongate = genome_dict[specie][strand_id][\"seq\"][\n",
    "                start_3:coordinates[0][0]\n",
    "            ].reverse_complement() # Get the 3' sequence\n",
    "\n",
    "\n",
    "    result_dict[\"nuc\"] = SeqRecord(seq = Seq(elongate), id = f\"{seq_id}-{cluster}\", description = \"\") \n",
    "\n",
    "    # def translate_frames(dna_sequence, specie, seq_id, length, utr, cluster)\n",
    "\n",
    "    result_dict[\"prot\"] = translate_frames(dna_sequence = elongate, specie = specie, seq_id = seq_id, length = custom_elongate_length, utr = side, cluster = cluster)\n",
    "\n",
    "    return result_dict\n",
    "    \n",
    "    \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def custom_scer_elongates(scer_id : str, specie : str, scer_length : int, scer_elongate : str or Seq, infos_dict, cluster, gff_dict, genome_dict, side):\n",
    "\n",
    "    coordinates = list()\n",
    "    result_dict = dict()\n",
    "\n",
    "    if side != \"Nter\" and side != \"Cter\":\n",
    "\n",
    "        raise ValueError(\"Side must be either Nter or Cter\")\n",
    "    \n",
    "    gff = gff_dict[specie].filter(\n",
    "\n",
    "        (pl.col(\"Type\") == \"CDS\") & ((pl.col(\"Name\") == scer_id) | (pl.col(\"Parent\") == scer_id))\n",
    "\n",
    "    )[[\"Start\",\"End\",\"Strand\",\"Seqid\"]] # Keep only necessary columns\n",
    "    \n",
    "    # Store datas necessary to compute the elongate sequence\n",
    "\n",
    "    strand = gff[0][\"Strand\"].to_list()[0] # + or -\n",
    "    strand_id = gff[0][\"Seqid\"].to_list()[0] # chromosome or scaffold id\n",
    "\n",
    "    for row in gff.iter_rows(named=True): \n",
    "\n",
    "        coordinates.append(sorted((int(row['Start'])-1, int(row['End'])-1))) # -1 for python indexing\n",
    "        \n",
    "    coordinates = sorted(coordinates, key=lambda x: x[0]) # Sort coordinates by start position\n",
    "\n",
    "    try:\n",
    "        scer_elongate = scer_elongate.replace(\"-\",\"\")\n",
    "    except AttributeError:\n",
    "        print(scer_id)\n",
    "        print(cluster)\n",
    "\n",
    "    if len(scer_elongate) != scer_length:\n",
    "\n",
    "        print(f\"Seq id : {scer_id}\")\n",
    "        print(f\"Scer length : {scer_length}\")\n",
    "        print(f\"Scer elongate length : {len(scer_elongate)}\")\n",
    "        print(f\"Cluster : {cluster}\")\n",
    "        raise ValueError(\"The length of the elongate sequence must be equal to the length of the Scer sequence\")\n",
    "\n",
    "    for subject_id, subject_length in infos_dict.items():\n",
    "\n",
    "        result_dict[subject_id] = dict()\n",
    "        custom_length = (scer_length-subject_length)*3\n",
    "\n",
    "        if strand == \"+\":\n",
    "\n",
    "            if side == \"Nter\":\n",
    "                \n",
    "                elongate = genome_dict[specie][strand_id][\"seq\"][\n",
    "\n",
    "                    coordinates[0][0]:coordinates[0][0] + custom_length # Get the 5' sequence\n",
    "\n",
    "                ] \n",
    "\n",
    "            elif side == \"Cter\": # Useless check but it's for the sake of clarity\n",
    "            \n",
    "\n",
    "                elongate = genome_dict[specie][strand_id][\"seq\"][\n",
    "                    \n",
    "                    coordinates[-1][1]-(custom_length):coordinates[-1][1]] # Get the 3' sequence\n",
    "\n",
    "    \n",
    "        # Reverse complement if the strand is negative, don't forget to reverse the coordinates\n",
    "        if strand == \"-\":\n",
    "\n",
    "            if side == \"Nter\":\n",
    "\n",
    "\n",
    "                elongate = genome_dict[specie][strand_id][\"seq\"][\n",
    "                    coordinates[-1][1]-(custom_length):coordinates[-1][1]\n",
    "                ].reverse_complement() # Get the 5' sequence\n",
    "            \n",
    "\n",
    "            elif side == \"Cter\":\n",
    "            \n",
    "                elongate = genome_dict[specie][strand_id][\"seq\"][\n",
    "                    coordinates[0][0]:coordinates[0][0] + (custom_length)\n",
    "                ].reverse_complement() # Get the 3' sequence\n",
    "\n",
    "\n",
    "        result_dict[subject_id][\"nuc\"] = SeqRecord( seq = Seq(elongate), id = f\"{scer_id}_{side}_{subject_id}\", description = \"\")\n",
    "        result_dict[subject_id][\"prot\"] = SeqRecord( seq = Seq(scer_elongate[:int(custom_length/3)]), id = f\"{scer_id}_{side}_{subject_id}\", description = \"\")\n",
    "    \n",
    "    return result_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import yaml\n",
    "# Load your species regex from the yaml file\n",
    "with open('/home/simon.herman/Bureau/Gits/Elongates/env.yaml', 'r') as f:\n",
    "    yaml_data = yaml.safe_load(f)\n",
    "    species_order = yaml_data['Species_order']['Scer']\n",
    "\n",
    "seuil = 4\n",
    "\n",
    "Nter_scer_conditions = ((pl.col(\"species\") == \"Scer_NCBI\") & (abs(pl.col(\"max_Nter\") - pl.col(\"Nter_nb_aa\")) < seuil) & (pl.col(\"max_Nter\") >= 10))\n",
    "Cter_scer_conditions = ((pl.col(\"species\") == \"Scer_NCBI\") & (abs(pl.col(\"max_Cter\") - pl.col(\"Cter_nb_aa\")) < seuil) & (pl.col(\"max_Cter\") >= 10))\n",
    "test = elongates.filter(Nter_scer_conditions | Cter_scer_conditions)[\"cluster_name\"].unique().to_list()\n",
    "\n",
    "nter_clusters = []\n",
    "cter_clusters = []\n",
    "i = 0\n",
    "\n",
    "for cluster,sequences in elongates.filter(pl.col(\"cluster_name\").is_in(test)).groupby(\"cluster_name\"):\n",
    "\n",
    "    species = sequences[\"species\"].to_list()\n",
    "\n",
    "    if len(species) < 3:\n",
    "            \n",
    "        continue\n",
    "\n",
    "\n",
    "    species_ordered = [ s for s in species_order if s in species]\n",
    "    \n",
    "    if species_ordered[-1] != \"Scer_NCBI\" and species.count(species_ordered[-1]) > 1:\n",
    "            \n",
    "        continue\n",
    "\n",
    "    if sequences.filter(pl.col(\"species\") == species_ordered[-1])[\"Nter_nb_aa\"].to_list()[0] < seuil and sequences[\"max_Nter\"].max() >= 10:\n",
    "    \n",
    "        nter_clusters.append(cluster)\n",
    "\n",
    "    if sequences.filter(pl.col(\"species\") == species_ordered[-1])[\"Cter_nb_aa\"].to_list()[0] < seuil and sequences[\"max_Cter\"].max() >= 10:\n",
    "    \n",
    "        cter_clusters.append(cluster)\n",
    "\n",
    "\n",
    "\n",
    "full_filtered_Nter = elongates.filter(pl.col(\"cluster_name\").is_in(nter_clusters))\n",
    "full_filtered_Cter = elongates.filter(pl.col(\"cluster_name\").is_in(cter_clusters))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gff3_parser\n",
    "from utils.files import multifasta_to_dict\n",
    "gff_dict = dict()\n",
    "genome_dict = dict()\n",
    "\n",
    "for specie in species_order:\n",
    "\n",
    "    gff_dict[specie] = pl.from_pandas(gff3_parser.parse_gff3(f\"input/{specie}.gff\", parse_attributes = True, verbose = False))\n",
    "    genome_dict[specie] = multifasta_to_dict(f\"input/{specie}.fna\", genome = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'prot'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-64-9d0535ac02c3>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     67\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     68\u001b[0m             \u001b[0mSeqIO\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mscer_elongates\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0msubject_sequence\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"nuc\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34mf\"{current_path}/local_align_files/{side}/{cluster}/{subject_sequence}/nucleotide/Scer_custom.fna\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"fasta\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 69\u001b[0;31m             \u001b[0mSeqIO\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mscer_elongates\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0msubject_sequence\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"prot\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34mf\"{current_path}/local_align_files/{side}/{cluster}/{subject_sequence}/protein/Scer_custom.faa\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"fasta\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     70\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'prot'"
     ]
    }
   ],
   "source": [
    "import os \n",
    "from Bio import SeqIO\n",
    "\n",
    "current_path = f\"/home/simon.herman/Bureau/Gits/Elongates/work/{cov}\"\n",
    "\n",
    "dataframes = {\n",
    "    \"Nter\": full_filtered_Nter,\n",
    "    \"Cter\": full_filtered_Cter,\n",
    "}\n",
    "\n",
    "os.mkdir(f\"{current_path}/local_align_files\")\n",
    "\n",
    "for side in [\"Nter\", \"Cter\"]:\n",
    "\n",
    "    df = dataframes[side]\n",
    "\n",
    "    os.mkdir(f\"{current_path}/local_align_files/{side}\")\n",
    "\n",
    "    for cluster, sequences in df.groupby(\"cluster_name\"):\n",
    "\n",
    "       \n",
    "\n",
    "\n",
    "        os.mkdir(f\"{current_path}/local_align_files/{side}/{cluster}\")\n",
    "\n",
    "        scer_length = sequences.filter(pl.col(\"species\") == \"Scer_NCBI\")[f\"{side}_nb_aa\"].max() # Maybe several Scer sequences in the cluster, we take the longest elongate\n",
    "        if scer_length < 10:\n",
    "            continue\n",
    "        scer_id = sequences.filter((pl.col(\"species\") == \"Scer_NCBI\") & (pl.col(f\"{side}_nb_aa\") == scer_length))[\"seq_id\"].to_list()[0]\n",
    "\n",
    "        ###################\n",
    "\n",
    "        # POURQUOI EST CE QUE CE CLUSTER ECHAPE AUX REGLES ??????\n",
    "\n",
    "        ###################\n",
    "\n",
    "        \n",
    "            \n",
    "        infos_for_scer_sequences = dict()\n",
    "        for sequence in sequences.iter_rows(named = True): \n",
    "\n",
    "            if scer_length - sequence[f\"{side}_nb_aa\"] >= 10:\n",
    "\n",
    "                os.mkdir(f\"{current_path}/local_align_files/{side}/{cluster}/{sequence['seq_id']}\")\n",
    "                os.mkdir(f\"{current_path}/local_align_files/{side}/{cluster}/{sequence['seq_id']}/nucleotide\")\n",
    "                os.mkdir(f\"{current_path}/local_align_files/{side}/{cluster}/{sequence['seq_id']}/protein\")\n",
    "\n",
    "\n",
    "                dict_ = custom_target_elongate(cluster, scer_length, sequence[\"seq_id\"], sequence[\"species\"], sequence[f\"{side}_nb_aa\"], side, genome_dict, gff_dict)\n",
    "\n",
    "                SeqIO.write(dict_[\"nuc\"], f\"{current_path}/local_align_files/{side}/{cluster}/{sequence['seq_id']}/nucleotide/{sequence['seq_id']}.fna\", \"fasta\")\n",
    "                for frame in dict_[\"prot\"].keys():\n",
    "                    SeqIO.write(dict_[\"prot\"][frame], f\"{current_path}/local_align_files/{side}/{cluster}/{sequence['seq_id']}/protein/{sequence['seq_id']}.faa\", \"fasta\")\n",
    "                infos_for_scer_sequences[sequence[\"seq_id\"]] = sequence[\"Nter_nb_aa\"] \n",
    "                \n",
    "        \n",
    "        scer_cds = sequences.filter(pl.col(\"seq_id\") == scer_id)\n",
    "        \n",
    "        scer_elongates = custom_scer_elongates(scer_id = scer_id, specie = \"Scer_NCBI\", scer_length = scer_length, \n",
    "                                                    scer_elongate = scer_cds[f\"{side}_elongate\"].to_list()[0], infos_dict = infos_for_scer_sequences, \n",
    "                                                    cluster = cluster, gff_dict = gff_dict, genome_dict = genome_dict, side = side)\n",
    "        if cluster == \"cluster_n1071\": \n",
    "            print(sequences[\"seq_id\",\"Nter_nb_aa\",\"max_Nter\"])\n",
    "            print(scer_elongates.keys())\n",
    "\n",
    "        for subject_sequence in scer_elongates.keys():\n",
    "\n",
    "            SeqIO.write(scer_elongates[subject_sequence][\"nuc\"], f\"{current_path}/local_align_files/{side}/{cluster}/{subject_sequence}/nucleotide/Scer_custom.fna\", \"fasta\")\n",
    "            SeqIO.write(scer_elongates[subject_sequence][\"prot\"], f\"{current_path}/local_align_files/{side}/{cluster}/{subject_sequence}/protein/Scer_custom.faa\", \"fasta\")\n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr > th,\n",
       ".dataframe > tbody > tr > td {\n",
       "  text-align: right;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (6, 8)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>seq_id</th><th>Cter_nb_aa</th><th>Nter_nb_aa</th><th>max_Nter</th><th>max_Cter</th><th>Nter_elongate_length</th><th>Cter_elongate_length</th><th>Cter_elongate</th></tr><tr><td>str</td><td>i64</td><td>i64</td><td>i64</td><td>i64</td><td>i64</td><td>i64</td><td>str</td></tr></thead><tbody><tr><td>&quot;Sarb_12G00060.…</td><td>0</td><td>0</td><td>72</td><td>11</td><td>0</td><td>0</td><td>null</td></tr><tr><td>&quot;Smik_12.6&quot;</td><td>0</td><td>72</td><td>72</td><td>11</td><td>72</td><td>0</td><td>null</td></tr><tr><td>&quot;rna-XM_0339118…</td><td>11</td><td>72</td><td>72</td><td>11</td><td>72</td><td>11</td><td>&quot;EEEYQFERHKL&quot;</td></tr><tr><td>&quot;rna-NM_0011818…</td><td>0</td><td>72</td><td>72</td><td>11</td><td>72</td><td>0</td><td>null</td></tr><tr><td>&quot;Sbay_10.18&quot;</td><td>0</td><td>72</td><td>72</td><td>11</td><td>72</td><td>0</td><td>null</td></tr><tr><td>&quot;Skud_12.17&quot;</td><td>0</td><td>72</td><td>72</td><td>11</td><td>72</td><td>0</td><td>null</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (6, 8)\n",
       "┌─────────────┬──────────┬────────────┬──────────┬──────────┬────────────┬────────────┬────────────┐\n",
       "│ seq_id      ┆ Cter_nb_ ┆ Nter_nb_aa ┆ max_Nter ┆ max_Cter ┆ Nter_elong ┆ Cter_elong ┆ Cter_elong │\n",
       "│ ---         ┆ aa       ┆ ---        ┆ ---      ┆ ---      ┆ ate_length ┆ ate_length ┆ ate        │\n",
       "│ str         ┆ ---      ┆ i64        ┆ i64      ┆ i64      ┆ ---        ┆ ---        ┆ ---        │\n",
       "│             ┆ i64      ┆            ┆          ┆          ┆ i64        ┆ i64        ┆ str        │\n",
       "╞═════════════╪══════════╪════════════╪══════════╪══════════╪════════════╪════════════╪════════════╡\n",
       "│ Sarb_12G000 ┆ 0        ┆ 0          ┆ 72       ┆ 11       ┆ 0          ┆ 0          ┆ null       │\n",
       "│ 60.1        ┆          ┆            ┆          ┆          ┆            ┆            ┆            │\n",
       "│ Smik_12.6   ┆ 0        ┆ 72         ┆ 72       ┆ 11       ┆ 72         ┆ 0          ┆ null       │\n",
       "│ rna-XM_0339 ┆ 11       ┆ 72         ┆ 72       ┆ 11       ┆ 72         ┆ 11         ┆ EEEYQFERHK │\n",
       "│ 11840.1     ┆          ┆            ┆          ┆          ┆            ┆            ┆ L          │\n",
       "│ rna-NM_0011 ┆ 0        ┆ 72         ┆ 72       ┆ 11       ┆ 72         ┆ 0          ┆ null       │\n",
       "│ 81874.2     ┆          ┆            ┆          ┆          ┆            ┆            ┆            │\n",
       "│ Sbay_10.18  ┆ 0        ┆ 72         ┆ 72       ┆ 11       ┆ 72         ┆ 0          ┆ null       │\n",
       "│ Skud_12.17  ┆ 0        ┆ 72         ┆ 72       ┆ 11       ┆ 72         ┆ 0          ┆ null       │\n",
       "└─────────────┴──────────┴────────────┴──────────┴──────────┴────────────┴────────────┴────────────┘"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "elongates.filter(pl.col(\"cluster_name\") == \"cluster_n3725\").select(\"seq_id\",\"Cter_nb_aa\",\"Nter_nb_aa\",\"max_Nter\",\"max_Cter\",\"Nter_elongate_length\",\"Cter_elongate_length\",\"Cter_elongate\")\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
